File: src/claw-generated/rpc/getSignaturesForAddress.rs
use crate::state::TransactionState;

pub async fn get_signatures_for_address(
    address: String,
    limit: usize,
    before: Option&lt;String&gt;,
    until: Option&lt;String&gt;,
) -&gt; Result&lt;Vec&lt;String&gt;, String&gt; {
    let transactions = TransactionState::get_transactions_for_address(&amp;address)
        .await
        .map_err(|e| format!("Error fetching transactions: {}", e))?;

    let mut signatures: Vec&lt;String&gt; = transactions
        .iter()
        .map(|tx| tx.signature.clone())
        .collect();

    signatures.sort_by(|a, b| {
        // Sort in descending order by timestamp
        let a_tx = TransactionState::get_transaction(a).await.unwrap();
        let b_tx = TransactionState::get_transaction(b).await.unwrap();
        b_tx.timestamp.cmp(&amp;a_tx.timestamp)
    });

    // Apply pagination
    let start = if let Some(before) = before {
        match signatures.binary_search(&amp;before) {
            Ok(idx) =&gt; idx + 1,
            Err(idx) =&gt; idx,
        }
    } else {
        0
    };

    let end = if let Some(until) = until {
        match signatures.binary_search(&amp;until) {
            Ok(idx) =&gt; idx + 1,
            Err(idx) =&gt; idx,
        }
    } else {
        signatures.len()
    };

    let paginated_signatures = &amp;signatures[start..start + limit.min(end - start)];
    Ok(paginated_signatures.to_vec())
}

File: src/claw-generated/rpc/getSignaturesForAddress.ts
import { PublicKey } from '@solana/web3.js';

interface GetSignaturesForAddressParams {
  address: string;
  limit?: number;
  before?: string;
  until?: string;
}

interface GetSignaturesForAddressResult {
  signatures: string[];
  before?: string;
  until?: string;
}

export async function getSignaturesForAddress(
  params: GetSignaturesForAddressParams
): Promise&lt;GetSignaturesForAddressResult&gt; {
  const { address, limit = 1000, before, until } = params;

  // Query the blockchain for signatures associated with the address
  const signatures = await getTransactionSignaturesForAddress(
    new PublicKey(address),
    { limit, before, until }
  );

  // Format the result with pagination info
  const result: GetSignaturesForAddressResult = {
    signatures: signatures.map((sig) => sig.signature),
    before: signatures.length === limit ? signatures[signatures.length - 1].signature : undefined,
    until: signatures.length === limit ? signatures[0].signature : undefined,
  };

  return result;
}

async function getTransactionSignaturesForAddress(
  address: PublicKey,
  { limit, before, until }: { limit?: number; before?: string; until?: string }
): Promise&lt;{ signature: string }[]&gt; {
  // Implement logic to fetch signatures from the blockchain
  // using the provided parameters
  // ...
}